File: src/hashers/CustomPasswordHasher.py
from django.contrib.auth.hashers import BasePasswordHasher
import hashlib
import os
from django.conf import settings


class CustomPasswordHasher(BasePasswordHasher):
    algorithm = "custom"

    def salt(self):
        # Use SECRET_KEY to generate deterministic salt
        secret_key = getattr(settings, "SECRET_KEY", "default_secret_key")
        return hashlib.sha256(secret_key.encode("utf-8")).hexdigest()[:16]

    def encode(self, password, salt):
        # Combine password and salt, then hash using SHA256
        return (
            hashlib.sha256((password + salt).encode("utf-8")).hexdigest() + "$" + salt
        )

    def verify(self, password, encoded):
        # Verify password by comparing hashes
        if "$" not in encoded:
            # Invalid encoded format, avoid ValueError
            return False
        parts = encoded.rsplit("$", 1)
        if len(parts) != 2 or not parts[0] or not parts[1]:
            # Either part is missing or empty, invalid format
            return False
        password_hash = parts[0]
        salt = parts[1]
        return (
            password_hash
            == hashlib.sha256((password + salt).encode("utf-8")).hexdigest()
        )

    def safe_summary(self, encoded):
        parts = encoded.rsplit("$", 1)
        salt = parts[1] if len(parts) == 2 else ""
        return {
            "algorithm": self.algorithm,
            "salt": salt,
        }

File: src/hashers/CustomPasswordHasher.php
<?php

use Dotenv\Dotenv;

class CustomPasswordHasher {
  private static function getSecretKey() {
    if (empty($_ENV['DJANGO_SECRET_KEY'])) {
      // Load environment variables from .env file
      $projectRootPath = dirname(__DIR__, 2); // Go up two levels from src/hashers
      $dotenv          = Dotenv::createImmutable($projectRootPath);
      $dotenv->load();
    }
    return $_ENV['DJANGO_SECRET_KEY'] ?? 'default_secret_key';
  }

  public static function getSalt() {
    $secretKey = self::getSecretKey();
    // Generate deterministic salt and convert it to hex
    return substr(hash('sha256', $secretKey), 0, 16);
  }

  public static function hash($password) {
    $salt           = self::getSalt();
    $hashedPassword = hash('sha256', $password . $salt); // Combine password and salt
    return $hashedPassword . '$' . $salt; // Return hashed password with salt
  }

  public static function verify($password, $encoded) {
    $parts = explode('$', $encoded, 2);
    if (count($parts) < 2) {
      // Invalid encoded format, avoid undefined array key warning
      return false;
    }
    list($passwordHash, $salt) = $parts;
    $hashedPassword            = hash('sha256', $password . $salt);
    return $passwordHash === $hashedPassword; // Compare hashes
  }

  public static function summary($encoded) {
    list($passwordHash, $salt) = explode('$', $encoded, 2);
    return ['algorithm' => 'custom', 'salt' => $salt];
  }
}
